### DISCUSSION TITLE
will you add support for tModloader?

### DISCUSSION THREAD

**Discussion by GamerKing122** [Pryaxis/TShock#1727] [General]
it would be good if later down the track you add support for tmodloader mods if they're willing to work with you so you can add permissions for certain mod commands

**Comment by QuiCM** [comment]
It's certainly something I'm interested in doing. We'll see how the chaos of 1.4/1.4mobile shakes things up

**Comment by hakusaro** [comment]
We've had a working version of TShock with TML at least once, but we never shipped it. It's way more possible than you think. OTAPI is really a killer feature.

**Comment by AxeelAnder** [comment]
To make this possible, we may need to cooperate with tml team. At least we need to separate a api lib then let them build it in tml just like the TerrariaHooks(MonoMod).

**Comment by inoribea** [comment]
could we base by this one?
https://github.com/DeathCradle/TML-TShock

**Reply by hakusaro** [reply]
Yo @DeathCradle you up?

**Reply by SignatureBeef** [reply]
@hakusaro yeh, thats just a POC. the concept could still work but i'm not sure how well with the latest tml updates (its been a while since i looked at this repo).

im actually in the process of rebuilding the otapi stack from the ground up on .net core/5 and have already started forward thinking about [tml](https://github.com/DeathCradle/Open-Terraria-API/blob/upcoming/OTAPI.Common/Remote.cs#L105). it will probably be a very similar approach and we can decide if tshock or tml wins in some scenarios but if the tml guys are on board with an integration like this it would make it much easier. i'm not far into the tml parts on the new stack and i don't recall the weird implications i came across last time either so i can't really promise much of it yet.

**Reply by tylerjwatson** [reply]
@DeathCradle If you want a hand with that let me know, I've been wanting to add a Javascript runtime into it for a while now so you can create modifications that don't need to be compiled, it'd be nice to just have an engine and expose cecil to a scripting engine

**Reply by SignatureBeef** [reply]
@tylerjwatson yeh sure, i love me some js so im happy to discuss it. i want otapi to head in a direction where it's easy for people who are not me to add modifications and to debug them, however there would still be a need for some c# modifications e.g. patches for monomod and hooks/c# code etc
the long term goal is to rip out the core to otapi and move it to a lib of its own too soo yeh.

**Comment by YSpoof** [comment]
That idea still up ?

**Reply by comdar16** [reply]
> That idea still up ?

Not for now i think.

**Reply by Hellcommander** [reply]
how about tmodloader 64 bit (has some different changes then main branch)

**Comment by TheCToTheY** [comment]
I have to say, I would love to be able to play large scale mods like Thorium or Calamity on tModLoader with TShock support. It would be so amazing. Any updates on this at all? With tmodloader 64 and whatnot?

**Reply by sgkoishi** [reply]
Not likely - it might be better to look for a management mod for tML.
Many features of TShock are about management and anti-cheating, which is not feasible to support arbitrary mod.

**Reply by TheCToTheY** [reply]
Do you have any suggestions for such a managament mod? That will allow things like /warp, /heal, /spawn, and set monster spawn rate /time etc?

**Reply by sgkoishi** [reply]
Not so familiar with mods but at least `HERO's Mod` could do some of these.
There are also some mods for developers that can directly run code to adjust whatever they want (only run trusted code), including setting mod-specific values.

For anti-cheating and permission management you probably need to write your own mod, because they typically only work for specific mod packs - if you add another mod, the new mechanisms will always be unmanaged or blocked.

**Comment by pmnlla** [comment]
Is this still a concept the devs are willing to take up? There's many QoL mods that I just can't get on vanilla terraria.

**Reply by sgkoishi** [reply]
Likely not, you can just use tModLoader without TShock.

**Reply by pmnlla** [reply]
Yes, but my problem is that tModLoader-server doesn't offer any interface outside of the limited console prompt, as far as I'm aware. This is the biggest strength of TShock

**Reply by sgkoishi** [reply]
What interface are you referring to? Do you mean things like SSC, commands, user management, anti-cheat, etc? Or things like REST API?

**Reply by pmnlla** [reply]
I'm referring to things like commands, user management, etc. Vanilla Terraria lacks those, and by extension tModLoader-server in and of itself as far as I'm aware..

**Reply by sgkoishi** [reply]
The user management are pretty much designed for ban and permission restrict, but it is impossible for TShock to know what mods will do, aka support tModLoader with mods - and it is not quite meaningful to support tModLoader without mods.

There are command system in vanilla and it is even localized! German players can use `/hilfe` instead of `/help`.

Mods like `HEROsMod` and `CheatSheet` might have what you want: allow register, have some buttons to toggle things that TShock can do, spawn item or boss, region protection, permission control for vanilla things, and even more powerful features that is only available on the client side.

**Reply by pmnlla** [reply]
I'll have a look at those two mods for sure. I do need to do a bit more research on terraria mods as well, so thanks for bringing those to my attention.

Circling back to your first point, now with that in mind it makes sense to not support tModloader, and while I argue that mods should implement permission systems like in Minecraft (see luckperms) at one point or another, it's not quite feasible to rework literally every existing mod for this purpose. I see where you're coming from and agree completely.

**Reply by avizmarlon** [reply]
@sgkoishi The things I value tshock most for are honestly: SSC and item ban. Do you know if there's a way to have this on tModLoader?

**Reply by sgkoishi** [reply]
> The things I value tshock most for are honestly: SSC and item ban. Do you know if there's a way to have this on tModLoader?

@avizmarlon For 1.3, there are several SSC mods, and we write our own item ban mod (which is just hard-coded several lines). Similar for 1.4, I just searched in the workshop and found 3 SSC mods (one of it provides item ban – and if it doesn't work, try to write your own).

**Comment by SavageTheUnicorn** [comment]
Curious if this is still a viable request.. Tshock runs fine on my ARM64 server but tmodloader definitely doesn't, atleast not 1.4.4. Makes me wonder if this could help get around that problem in the most complicated way possible lol

**Reply by PseudoDistant** [reply]
Just out of curiosity, would that ARM64 server be a Raspberry Pi?
That's exactly why I'm here lol

**Reply by pmnlla** [reply]
Could also be an ampere box. Either way is good.

**Reply by hakusaro** [reply]
https://github.com/MonoMod/MonoMod/issues/90 is the biggest issue towards getting any support because MonoMod rework doesn’t support arm at all and nobody seems to care despite me and several others offering thousands of dollars of support and hardware
